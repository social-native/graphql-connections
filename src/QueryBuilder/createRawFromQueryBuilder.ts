import {QueryBuilder, RawBuilder} from 'knex';

/**
 * Knex does not provide a createRawFromQueryBuilder, so this fills in what Knex does in:
 * https://github.com/tgriesser/knex/blob/887fb5392910ab00f491601ad83383d04b167173/src/util/make-knex.js#L29
 */
export default function createRawFromQueryBuilder(
    builder: QueryBuilder,
    rawSqlQuery: string,
    bindings?: any
) {
    const {
        client
    }: {
        client: IStolenClient;
    } = builder as any;

    const args = [rawSqlQuery, bindings].filter(arg => arg);

    return client.raw.apply(client, args as any);
}

interface IStolenClient {
    raw: RawBuilder;
}
