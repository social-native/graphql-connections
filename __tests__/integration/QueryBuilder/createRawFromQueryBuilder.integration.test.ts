import knex from 'knex';
import {test as testConfig} from '../../../knexfile.sqlite';
import createRawFromQueryBuilder from '../../../src/QueryBuilder/createRawFromQueryBuilder';
const knexClient = knex(testConfig);

describe('QueryBuilder', function() {
    describe('createRawFromQueryBuilder', function() {
        describe('given a raw query', function() {
            let result: knex.Raw;

            beforeAll(function() {
                result = createRawFromQueryBuilder(knexClient.queryBuilder(), '1');
            });

            it('produces a query segment', function() {
                expect(result.toString()).toMatchSnapshot();
            });

            it('can be used in a select', async function() {
                expect(
                    await knexClient
                        .from('mock')
                        .select(result)
                        .limit(1)
                ).toMatchSnapshot();
            });
        });

        describe('given a raw query and bindings', function() {
            let result: knex.Raw;

            beforeAll(function() {
                result = createRawFromQueryBuilder(
                    knexClient.queryBuilder(),
                    '(:sweet) as flavor',
                    {
                        sweet: 'sugar'
                    }
                );
            });

            it('produces an interpolated query segment', function() {
                expect(result.toString()).toMatchSnapshot();
            });

            it('can be used in a select', async function() {
                expect(
                    await knexClient
                        .from('mock')
                        .select(result)
                        .limit(1)
                ).toMatchSnapshot();
            });
        });
    });
});
