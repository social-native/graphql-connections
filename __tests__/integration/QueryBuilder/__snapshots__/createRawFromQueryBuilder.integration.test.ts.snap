// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryBuilder createRawFromQueryBuilder given a raw query and bindings can be used in a select 1`] = `
Array [
  Object {
    "flavor": "sugar",
  },
]
`;

exports[`QueryBuilder createRawFromQueryBuilder given a raw query and bindings produces an interpolated query segment 1`] = `"('sugar') as flavor"`;

exports[`QueryBuilder createRawFromQueryBuilder given a raw query can be used in a select 1`] = `
Array [
  Object {
    "1": 1,
  },
]
`;

exports[`QueryBuilder createRawFromQueryBuilder given a raw query produces a query segment 1`] = `"1"`;
